### PR TITLE
HHH-17210 - Expose custom JFR events

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-17210

https://hibernate.atlassian.net/browse/HHH-17270

### CODE REVIEW

**Review by sebersole — COMMENTED** [review]
I kind of think these "session events" should include the sessionIdentifier.

For the "Connection acquisition" event, maybe how long it took?

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 1]
```diff
@@ -602,14 +602,16 @@ public JdbcConnectionAccess getJdbcConnectionAccess() {
 			if ( ! fastSessionServices.requiresMultiTenantConnectionProvider ) {
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:602]: The changes in this block are what I was referring to about now-eager generation of the session-identifier

File: hibernate-core/src/main/java/org/hibernate/action/internal/EntityInsertAction.java [hunk 2]
```diff
@@ -161,11 +164,23 @@ protected void putCacheIfNecessary() {
...
+		final EntityDataAccess cacheAccessStrategy = persister.getCacheAccessStrategy();
+		boolean insert = false;
 		try {
 			session.getEventListenerManager().cachePutStart();
-			return persister.getCacheAccessStrategy().insert( session, ck, cacheEntry, version );
+			insert = cacheAccessStrategy.insert( session, ck, cacheEntry, version );
+			return insert;
 		}
 		finally {
+			JfrEventManager.commitCachePutEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/action/internal/EntityInsertAction.java:176]: Here is another case of now-eager `sessionIdentifier` generation.

File: hibernate-core/src/main/java/org/hibernate/engine/internal/CacheHelper.java [hunk 3]
```diff
@@ -28,14 +32,58 @@ private CacheHelper() {
...
 			CachedDomainDataAccess cacheAccess) {
 		final SessionEventListenerManager eventListenerManager = session.getEventListenerManager();
 		Object cachedValue = null;
 		eventListenerManager.cacheGetStart();
+		final CacheGetEvent cacheGetEvent = JfrEventManager.beginCacheGetEvent();
 		try {
 			cachedValue = cacheAccess.get( session, cacheKey );
 		}
 		finally {
+			JfrEventManager.commitCacheGetEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/CacheHelper.java:80]: Here is another case of now-eager `sessionIdentifier` generation.

File: hibernate-core/src/main/java/org/hibernate/engine/internal/NaturalIdResolutionsImpl.java [hunk 4]
```diff
@@ -268,32 +270,63 @@ private void manageSharedResolution(
...
-							cacheAccess.getRegion().getName()
+					if ( put && statistics.isStatisticsEnabled() ) {
+						statistics.naturalIdCachePut(
+								rootEntityDescriptor.getNavigableRole(),
+								cacheAccess.getRegion().getName()
+						);
+					}
+				}
+				finally {
+					JfrEventManager.commitCachePutEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/NaturalIdResolutionsImpl.java:295]: Here is another case of now-eager `sessionIdentifier` generation.

File: hibernate-core/src/main/java/org/hibernate/engine/internal/NaturalIdResolutionsImpl.java [hunk 5]
```diff
@@ -326,29 +359,59 @@ private void manageSharedResolution(
...
+
+									if ( putAfterUpdate && statistics.isStatisticsEnabled() ) {
+										statistics.naturalIdCachePut(
+												rootEntityDescriptor.getNavigableRole(),
+												cacheAccess.getRegion().getName()
+										);
+									}
+								}
+								finally {
+									JfrEventManager.commitCachePutEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/NaturalIdResolutionsImpl.java:407]: Here is another case of now-eager `sessionIdentifier` generation.

File: hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+	public static SessionOpenEvent beginSessionOpenEvent() {
+		final SessionOpenEvent sessionOpenEvent = new SessionOpenEvent();
+		if ( sessionOpenEvent.isEnabled() ) {
+			sessionOpenEvent.begin();
+		}
+		return sessionOpenEvent;
+	}
+
+	public static void commitSessionOpenEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java:42]: Would the name `#completeXYZ` be better to correlate better with `#beginXYZ`?  

JFR calls the end operation commit, so I understand that naming too.

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 7]
```diff
@@ -602,14 +602,16 @@ public JdbcConnectionAccess getJdbcConnectionAccess() {
...
 				jdbcConnectionAccess = new NonContextualJdbcConnectionAccess(
 						getEventListenerManager(),
-						fastSessionServices.connectionProvider
+						fastSessionServices.connectionProvider,
+						getSessionIdentifier()
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:606]: Here is another case of now-eager `sessionIdentifier` generation.

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 8]
```diff
@@ -602,14 +602,16 @@ public JdbcConnectionAccess getJdbcConnectionAccess() {
...
 				);
 			}
 			else {
 				jdbcConnectionAccess = new ContextualJdbcConnectionAccess(
 						getTenantIdentifier(),
 						getEventListenerManager(),
-						fastSessionServices.multiTenantConnectionProvider
+						fastSessionServices.multiTenantConnectionProvider,
+						getSessionIdentifier()
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:614]: Here is another case of now-eager `sessionIdentifier` generation.

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java [hunk 9]
```diff
@@ -994,33 +996,56 @@ private void putInCache(
...
+				update = cacheAccess.update(
+						session,
+						cacheKey,
+						concreteDescriptor.getCacheEntryStructure().structure( cacheEntry ),
+						version,
+						version
+				);
+			}
+			finally {
+				JfrEventManager.commitCachePutEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java:1011]: Here is another case of now-eager sessionIdentifier generation.

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java [hunk 10]
```diff
@@ -994,33 +996,56 @@ private void putInCache(
...
 						cacheKey,
 						concreteDescriptor.getCacheEntryStructure().structure( cacheEntry ),
 						version,
 						//useMinimalPuts( session, entityEntry )
 						false
 				);
-
+			}
+			finally {
+				JfrEventManager.commitCachePutEvent(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java:1037]: Here is another case of now-eager sessionIdentifier generation.

File: hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java [hunk 11]
```diff
@@ -297,16 +301,27 @@ private static void addCollectionToCache(
 								&& session.getCacheMode()!= CacheMode.REFRESH
 				);
 
+
+			}
+			finally {
+				JfrEventManager.commitCachePutEvent(
+						cachePutEvent,
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java:308]: Here is another case of now-eager sessionIdentifier generation.

**Review by sebersole — COMMENTED** [review]
At the moment (b4 this work), generating the `sessionIdentifier` was very lazily protected (presumably for perf because it generates a UUID).

With these changes, it is generated much more frequently.  I notated the places I saw this happen with PR comments.  The one I especially questioned was the code to create `JdbcConnectionAccess`, which will effectively happen for every `Session` negating the benefit (if one) to delay generation of the `sessionIdentifier`.  We need to decide if that laziness is truly worth it and then either -

1. Generate `sessionIdentifier` up-front and make it final
2. Pass Session (or some form of delayable `sessionIdentifier` access) in these cases

File: hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java (hunk 6 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java:42]: I agree `completeXYZ` is better

File: hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+
+	public static void completeJdbcPreparedStatementExecutionEvent(
+			JdbcPreparedStatementExecutionEvent jdbcPreparedStatementExecutionEvent,
+			PreparedStatement statement) {
+		if ( jdbcPreparedStatementExecutionEvent.isEnabled() ) {
+			jdbcPreparedStatementExecutionEvent.end();
+			if ( jdbcPreparedStatementExecutionEvent.shouldCommit() ) {
+				jdbcPreparedStatementExecutionEvent.executionTime = getExecutionTime(
+						jdbcPreparedStatementExecutionEvent.startedAt );
+				jdbcPreparedStatementExecutionEvent.sql = statement.toString();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java:176]: This is not safe i.e. `toString` is not guaranteed to return the SQL string. Would be good if you could fix the callsites to use the variant that passes the SQL string directly.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java:178]: this method is called where we do not have access to the SQL String, for all the other cases there is the
```
completeJdbcPreparedStatementExecutionEvent(
			JdbcPreparedStatementExecutionEvent jdbcPreparedStatementExecutionEvent,
			String preparedStatementSql) {
```
method.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java:180]: @beikov see `org.hibernate.engine.jdbc.internal.ResultSetReturnImpl`

- `extract(PreparedStatement statement)`
-  `extract(CallableStatement callableStatement)`
-  `execute(PreparedStatement statement)`
-  `executeUpdate(PreparedStatement statement)`

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java:180]: they are all deprecated and it seems we are not using them anymore

**Pull request closed without merging**